### PR TITLE
feat: middleware and virtual routes

### PR DESCRIPTION
**Pull request by lilnasy**
## Changes

- Allows middleware to run for 404 responses without needing a 404.astro in the project.
- Fixes https://github.com/withastro/astro/issues/8244
- Fixes https://github.com/withastro/astro/issues/8226

## Testing
- Adapted from https://github.com/withastro/astro/pull/10185/files

## Docs
Mentions of `404.astro` or spread routes needing to exist to run middleware should be removed.

### CODE REVIEW

File: packages/astro/test/astro-dev-headers.test.js [hunk 1]
```diff
@@ -31,10 +31,10 @@ describe('Astro dev headers', () => {
 			assert.equal(Object.fromEntries(result.headers)['x-astro'], headers['x-astro']);
 		});
 
-		it('does not return custom headers for invalid URLs', async () => {
+		it('returns custom headers in the default 404 response', async () => {
 			const result = await fixture.fetch('/bad-url');
 			assert.equal(result.status, 404);
-			assert.equal(Object.fromEntries(result.headers).hasOwnProperty('x-astro'), false);
+			assert.equal(Object.fromEntries(result.headers).hasOwnProperty('x-astro'), true);
 		});
```
Comment (lilnasy) [inline packages/astro/test/astro-dev-headers.test.js:38]: The test checks that configured headers added in `config.server.headers` are NOT added in the 404 responses.

This behavior was limited to astro's default 404 page - if the user added a custom `404.astro`, `server.headers` are still added.

The new behavior is that `server.headers` apply to all astro-generated responses - the distinction between astro's default 404 page and user's `404.astro` is gone.

**Comment by matthewp** [comment]
lgtm, still in draft though and missing a changeset.

**Comment by lilnasy** [comment]
Manually adding `semver: minor` because label workflow doesn't work on forks.

**Review by OliverSpeir — CHANGES_REQUESTED** [review]
blocking per request

**Review by OliverSpeir — APPROVED** [review]

File: .changeset/brown-pets-clean.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+"astro": minor
+---
+
+The astro middleware now runs when a matching page or endpoint is not found. Previously, a `pages/404.astro` or `pages/[...catch-all].astro` route had to match to allow middleware. This is now not necessary.
+
+Note that some adapters may need to update to let Astro's SSR code handle the not-found case.
```
Comment (ematipico) [inline .changeset/brown-pets-clean.md:7]: nit: "to update.." what? It feels like we should hint at something or maybe rephrase the sentence. 

It would be great if the changelog actually explains what an adapter maintainer should update in order to "fix" this.

File: packages/astro/src/core/routing/astro-designed-error-pages.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import type { ManifestData } from "../../@types/astro.js";
+
+export function ensure404Route(manifest: ManifestData) {
+	if (!manifest.routes.some(route => route.route === '/404')) {
+		manifest.routes.push({
+			component: 'astro-default-404',
```
Comment (ematipico) [inline packages/astro/src/core/routing/astro-designed-error-pages.ts:6]: Let's export this `'astro-default-404'` as a constant, so we use this constant in all the files that check for this string.

File: packages/astro/src/vite-plugin-astro-server/pipeline.ts [hunk 4]
```diff
@@ -133,6 +134,9 @@ export class DevPipeline extends Pipeline {
 
 	async preload(filePath: URL) {
 		const { loader } = this;
+		if (filePath.href === new URL('astro-default-404', this.config.root).href) {
+			return { default: default404Page } as any as ComponentInstance
```
Comment (ematipico) [inline packages/astro/src/vite-plugin-astro-server/pipeline.ts:138]: Should we use `unknown` instead of `any`? Or is the `as any` necessary?

File: packages/astro/src/vite-plugin-astro-server/response.ts [hunk 5]
```diff
@@ -23,6 +23,18 @@ export async function handle404Response(
...
+	{ pathname }: { pathname: string }
+) {
+	return new Response(notFoundTemplate({
+		statusCode: 404,
+		title: 'Not found',
+		tabTitle: '404: Not Found',
+		pathname,
+	}), { status: 404, headers: { 'Content-Type': 'text/html; charset=utf-8' } });
+}
+default404Page.isAstroComponentFactory = true;
```
Comment (ematipico) [inline packages/astro/src/vite-plugin-astro-server/response.ts:36]: Let's add a comment that explains what this is doing and why

File: packages/astro/src/vite-plugin-astro-server/route.ts [hunk 6]
```diff
@@ -95,6 +95,19 @@ export async function matchRoute(
 	}
 
 	const custom404 = getCustom404Route(manifestData);
+	
+	if (custom404 && custom404.component === 'astro-default-404') {
+		const component: any = {
```
Comment (ematipico) [inline packages/astro/src/vite-plugin-astro-server/route.ts:100]: Can we avoid the `any` and use `@ts-expect-error`, so we can add a comment that explains why we're doing this?

**Comment by ematipico** [comment]
@lilnasy What about the documentation? The template is empty. We should have a PR for that, right?

File: .changeset/brown-pets-clean.md (hunk 2 above)
Comment (lilnasy) [reply .changeset/brown-pets-clean.md:7]: I feel it adequately explains the change necessary.

Adapters may look at the list of routes given by astro to create their platform-specific routing file. And the platform may then run SSR code only for routes known to be on-demand rendered. The website would then fallback to a platform-designed 404 page. SST, vercel and cloudflare (in one routing strategy) work like this.

File: packages/astro/src/vite-plugin-astro-server/pipeline.ts (hunk 4 above)
Comment (lilnasy) [reply packages/astro/src/vite-plugin-astro-server/pipeline.ts:138]: There wouldn't be any difference between the two in terms of safety.

File: packages/astro/src/vite-plugin-astro-server/response.ts (hunk 5 above)
Comment (lilnasy) [reply packages/astro/src/vite-plugin-astro-server/response.ts:36]:
```suggestion
// mark the function as an AstroComponentFactory for the rendering internals
default404Page.isAstroComponentFactory = true;
```

**Comment by lilnasy** [comment]
Not necessarily. We don't explicitly document that middleware won't run in this case. We should also hold explicitly saying that will , until we know it works this way across adapters.

File: packages/astro/src/vite-plugin-astro-server/route.ts (hunk 6 above)
Comment (lilnasy) [reply packages/astro/src/vite-plugin-astro-server/route.ts:100]: The types around these parts are wishy washy, so you'd see a lot of any's. In this case, the correct type error'd. I'll  take another look.

**Comment by ematipico** [comment]
> Not necessarily. We don't explicitly document that middleware won't run in this case. 

Sorry for being pedantic, but I am a bit concerned by this kind of approach: 
1. Can you update the PR template and explain why docs don't need to be updated? 
2. Since this is a** feature**, we should actually document this **new** behaviour.

> We should also hold explicitly saying that will , until we know it works this way across adapters.

And how do we plan to do that? The PR doesn't say anything about it.

**Comment by lilnasy** [comment]
updated the description

**Review by ematipico — APPROVED** [review]
Thank you, Arsh, for addressing my concerns! Another `minor`ready for the next release. Let's wait for docs :)

File: .changeset/brown-pets-clean.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+---
+"astro": minor
+---
+
+The astro middleware now runs when a matching page or endpoint is not found. Previously, a `pages/404.astro` or `pages/[...catch-all].astro` route had to match to allow middleware. This is now not necessary.
```
Comment (sarah11918) [inline .changeset/brown-pets-clean.md:5]:
```suggestion
Allows middleware to run when a matching page or endpoint is not found. Previously, a `pages/404.astro` or `pages/[...catch-all].astro` route had to match to allow middleware. This is now not necessary.
```

**Review by sarah11918 — APPROVED** [review]
Tiny suggestion from docs!

File: .changeset/brown-pets-clean.md (hunk 7 above)
Comment (lilnasy) [reply .changeset/brown-pets-clean.md:5]: Looks great!

**Comment by ematipico** [comment]
We should update the documentation: https://docs.astro.build/en/guides/middleware/

We need to mention that now the middleware will run on 404 pages
